File: core/src/main/java/com/jd/core/base/BaseFragment.java
package com.jd.core.base;


import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.gyf.barlibrary.ImmersionBar;
import com.jd.core.utils.AppLog;

import butterknife.ButterKnife;
import butterknife.Unbinder;

public abstract class BaseFragment extends Fragment {
    protected Context mContext;
    private Unbinder unBinder;
    public ImmersionBar mImmersionBar;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.mContext = context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(getLayout(), container, false);
        unBinder = ButterKnife.bind(this , view);
        //沉浸式状态栏
        initImmersionBar();
        initTitle();

        return view ;
    }

    /**
     * 返回view
     * @param view
     * @param savedInstanceState
     */
    @Override
    public void onViewCreated(@NonNull View view,@Nullable Bundle savedInstanceState) {
        super.onViewCreated(view,savedInstanceState);
        init(view);
        initView();
    }


    @Override
    public void onResume() {
        super.onResume();
        AppLog.i("当前运行的fragment:" + getClass().getName());
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (unBinder != null) {
            unBinder.unbind();
        }
        if (mImmersionBar != null){
            mImmersionBar.destroy();
        }
    }


    private void initImmersionBar() {
        if (mImmersionBar == null){
            mImmersionBar = ImmersionBar.with(this);
            mImmersionBar.init();
        }
    }

    protected void init(View view){};
    /**
     * 获取当前Activity的UI布局
     *
     * @return 布局id
     */
    protected abstract int getLayout();

    /**
     * 初始化标题
     */
    protected abstract void initTitle();

    /**
     * 初始化数据
     */
    protected abstract void initView();

}

File: JDApp-Kotlin/lib/core/src/main/java/com/jd/core/mvvm/v/BaseFragment.kt
package com.jd.core.mvvm.v


import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import androidx.lifecycle.ViewModelProvider

import com.gyf.barlibrary.ImmersionBar
import com.jd.core.R
import com.jd.core.mvvm.vm.BaseViewModel
import com.jd.core.view.NavigationBar


abstract class BaseFragment : Fragment() {
    private var mContext: Context? = null
    private var mImmersionBar: ImmersionBar? = null

    //导航
    lateinit var navigationBar: NavigationBar

    //contentView
    private lateinit var mContentView: View
    /***************************** 抽象方法  */
    /**
     * 获取当前Activity的UI布局
     *
     * @return 布局id
     */
    protected abstract fun layoutView() : View


    override fun onAttach(context: Context) {
        super.onAttach(context)
        this.mContext = context
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.base_layout, container, false)
        //导航
        this.navigationBar = view.findViewById(R.id.navigation_bar)
        //隐藏导航
        if (this.preferredNavigationBarHidden()) {
            this.navigationBar.visibility = View.GONE
        }

        val layoutView = this.layoutView()
        //添加子contentView
        val rl = view.findViewById<RelativeLayout>(R.id.base_content)
        rl.addView(layoutView)
        this.mContentView = layoutView
        //沉浸式状态栏
        initImmersionBar()
        return view
    }

    private fun initImmersionBar() {
        if (mImmersionBar == null) {
            mImmersionBar = ImmersionBar.with(this)
            mImmersionBar!!.init()
        }
    }

    /**
     * 获取子ViewModel
     */
    fun findViewModel(clazz:Class<out BaseViewModel>) : BaseViewModel {
        return ViewModelProvider(this)[clazz]
    }

    /**
     * 返回view
     * @param view
     * @param savedInstanceState
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView(view)
    }

    override fun onDestroy() {
        super.onDestroy()
        if (mImmersionBar != null) {
            mImmersionBar!!.destroy()
        }
    }

    /**
     * 隐藏导航
     */
    protected open fun preferredNavigationBarHidden(): Boolean {
        return false
    }

    /**
     * 初始化数据
     */
    protected abstract fun initView(view: View)

    /**
     * 懒加载数据
     */
    protected  abstract  fun loadData()

}
